crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -42,74 +252,151 @@ circuit_cache_key!(local ExchangeId<T>(usize => Arc<Exchange<T>>));
-    /// Callback invoked when all `npeers` mailboxes are available.
-    sender_callbacks: Vec<OnceCell<Box<dyn Fn() + Send + Sync>>>,
+    inner: Arc<Inner>,
+    /// `npeers^2` mailboxes, clients, and servers, one for each sender/receiver
+    /// pair.  Each mailbox is accessed by exactly two threads, so contention is
+    /// low.
+    mailboxes: Arc<Vec<Mutex<Option<T>>>>,
 }
 
 impl<T> Exchange<T>
 where
-    T: Send + 'static,
+    T: Clone + Send + Encode + Decode + 'static,
 {
     /// Create a new exchange operator for `npeers` communicating threads.
-    fn new(npeers: usize) -> Self {
-        Self {
-            npeers,
-            mailboxes: (0..npeers * npeers).map(|_| Mutex::new(None)).collect(),
-            receiver_counters: (0..npeers)
-                .map(|_| CachePadded::new(AtomicUsize::new(0)))
-                .collect(),
-            receiver_callbacks: (0..npeers).map(|_| OnceCell::new()).collect(),
-            sender_counters: (0..npeers)
-                .map(|_| CachePadded::new(AtomicUsize::new(npeers)))
-                .collect(),
-            sender_callbacks: (0..npeers).map(|_| OnceCell::new()).collect(),
-        }
+    fn new(
+        runtime: &Runtime,
+        tokio: TokioHandle,
+        exchange_id: ExchangeId,
+        clients: Arc<Clients>,
+        directory: ExchangeDirectory,
+    ) -> Self {
+        let npeers = runtime.num_workers();
+        let mailboxes: Arc<Vec<Mutex<Option<T>>>> =
+            Arc::new((0..npeers * npeers).map(|_| Mutex::new(None)).collect());
+        let mailboxes2: Arc<Vec<Mutex<Option<T>>>> = mailboxes.clone();
+        let deliver = move |data: Vec<u8>, sender, receiver| {
+            let index: usize = sender * npeers + receiver;
+            let data = decode_from_slice(&data, bincode::config::standard())
+                .unwrap()
+                .0;
+            let mut mailbox = mailboxes2[index].lock().unwrap();
+            assert!((*mailbox).is_none());
+            *mailbox = Some(data);
+        };
+
+        let inner = Arc::new(Inner::new(runtime, tokio, exchange_id, deliver, clients));
+        directory
+            .write()
+            .unwrap()
+            .entry(exchange_id)
+            .and_modify(|_| panic!())
+            .or_insert(inner.clone());
+        Self { inner, mailboxes }
+    }
+
+    /// Returns a reference to a mailbox for the sender/receiver pair.
+    fn mailbox(&self, sender: usize, receiver: usize) -> &Mutex<Option<T>> {
+        &self.mailboxes[self.inner.mailbox_index(sender, receiver)]
     }
 
     /// Create a new `Exchange` instance if an instance with the same id
     /// (created by another thread) does not yet exist within `runtime`.
     /// The number of peers will be set to `runtime.num_workers()`.
-    pub(crate) fn with_runtime(runtime: &Runtime, exchange_id: usize) -> Arc<Self> {
+    pub(crate) fn with_runtime(runtime: &Runtime, exchange_id: ExchangeId) -> Arc<Self> {
```

Comment:
I would love a diagram (probably in the README) showing how the data structures are distributed across processes.